sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala

```diff
@@ -242,15 +242,24 @@ abstract class BaseSessionStateBuilder(
    * Note: this depends on `catalog` and `experimentalMethods` fields.
    */
   protected def optimizer: Optimizer = {
-    new SparkOptimizer(catalogManager, catalog, experimentalMethods) {
-      override def earlyScanPushDownRules: Seq[Rule[LogicalPlan]] =
-        super.earlyScanPushDownRules ++ customEarlyScanPushDownRules
+    val optimizerClassName = conf.getConfString("spark.sql.customized.optimizer", "")
+    if (optimizerClassName.nonEmpty && !optimizerClassName.equals("")) {
+      val catalogManagerTypeName = Class.forName("org.apache.spark.sql.connector.catalog.CatalogManager")
+      val sessionCatalogTypeName = Class.forName("org.apache.spark.sql.catalyst.catalog.SessionCatalog")
+      val experimentalMethodsTypeName = Class.forName("org.apache.spark.sql.ExperimentalMethods")
+      val optimizerInstance = Class.forName(optimizerClassName).getConstructor(catalogManagerTypeName, sessionCatalogTypeName, experimentalMethodsTypeName).newInstance(catalogManager, catalog, experimentalMethods)
```

Comment:
file line length exceeds 100 characters on line 247,248,250 and 304

https://github.com/databricks/scala-style-guide#linelength